### PR TITLE
Fix `LruCache` corruption bug with a `size_callback` that can return 0

### LINKED ISSUE #11436: RoomInitialSyncRestServlet internal server error with "AssertionError synapse.util.linked_list in move_after"
### Description

A host started throwing `AssertionError synapse.util.linked_list in move_after` in `GET' uri='/_matrix/client/r0/rooms/!abcdefgh%3Adomain.tld/initialSync?limit=20&access_token=<redacted>`.

The particular client calling that URL had been fine for a long time. It seems the error started happening at or some time after a restart of Synapse and the two workers (synchrotron + initial synchrotron).

After restarting everything again, the error disappeared.

### Version information

- **Homeserver**: EMS customer, please pm for details and times
- **Version**: v1.47.1
- **Install method**: Docker

### PR DESCRIPTION
**Pull request by squahtx**
When all entries in an `LruCache` have a size of 0 according to the
provided `size_callback`, and `drop_from_cache` is called on a cache
node, the node would be unlinked from the LRU linked list but remain in
the cache dictionary. An assertion would be later be tripped due to the
inconsistency.

Avoid unintentionally calling `__len__` and use a strict `is None`
check instead when unwrapping the weak reference.

Fixes #11436

### CODE REVIEW

**Comment by DMRobertson** [comment]
> Use a strict is None check instead when unwrapping the weak reference.

something something std::option::Optional

File: synapse/util/caches/lrucache.py [hunk 1]
```diff
@@ -270,7 +270,7 @@ def drop_from_cache(self) -> None:
         removed from all lists.
         """
         cache = self._cache()
-        if not cache or not cache.pop(self.key, None):
+        if cache is None or not cache.pop(self.key, None):
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/util/caches/lrucache.py:273]: I think the 2nd clause is to check if the entry was not in the cache. `is None` is more correct but still does the wrong thing (in spirit) if the cache value is `None`. Luckily `drop_from_lists()` is idempotent so it's fine functionally.

[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/util/caches/lrucache.py]: I realised I could use the `.pop(..., sentinel) is sentinel` pattern seen elsewhere in the file.

File: synapse/util/caches/lrucache.py [hunk 2]
```diff
@@ -271,7 +271,10 @@ def drop_from_cache(self) -> None:
         removed from all lists.
         """
         cache = self._cache()
-        if cache is None or cache.pop(self.key, None) is None:
+        if (
+            cache is None
+            or cache.pop(self.key, _Sentinel.sentinel) is _Sentinel.sentinel
```
Comment (DMRobertson) [inline synapse/util/caches/lrucache.py:276]: So that we can store `None` as a value in the cache?

**Review by DMRobertson — APPROVED** [review]
Could maybe mention in the changelog that requests which hit this would fail consistently until Synapse was restarted? I'll leave that to your discretion.

Otherwise LGTM

File: synapse/util/caches/lrucache.py (hunk 2 above)
Comment (squahtx) [reply synapse/util/caches/lrucache.py:276]: That was the thinking, yes.

Although `drop_from_lists` is idempotent so it's functionally the same

**Comment by squahtx** [comment]
> Could maybe mention in the changelog that requests which hit this would fail consistently until Synapse was restarted? I'll leave that to your discretion.
> 
> Otherwise LGTM

Good idea, I've reworded it.

**Review by DMRobertson — APPROVED** [review]

**Pull request merged**